### PR TITLE
Add TypeScript

### PR DESCRIPTION
**Pull request by alifa20**
Introduced packages: 
- Rollup 
- @emotion/react
- ts-jest
- react version 17 upgrade
- yalc for checking the published package

Breaking changes:
- `@emotion/react` needs to be installed.
- node version  has to be the latest v12.21.0

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,26 +33,32 @@
...
   },
   "peerDependencies": {
-    "@emotion/core": "^10.0.28",
-    "prop-types": "^15.7.2",
-    "react": "^16.13.0",
-    "react-dom": "^16.13.0"
+    "@emotion/react": "11.1.5",
+    "prop-types": "15.7.2",
+    "react": "17.0.2",
+    "react-dom": "17.0.2"
```
[the remark this answers is not in the document]
Comment (moroshko) [reply package.json:39]: Let's start writing a migration guide to 2.0.
Could you add a new `MIGRATING_TO_2.0.md` in the root of the project?
Here is what comes to mind:
* If you're using an older version of React, you'll have to upgrade to React 17.
* You'll have to upgrade to Emotion 11. The main thing to notice is that you'll need to uninstall `@emotion/core` and install `@emotion/react` instead. More details [here](https://emotion.sh/docs/emotion-11).

Please add more stuff that you can think of to this list!

File: package-lock.json [hunk 2]
```diff
@@ -1,14315 +1,8 @@
 {
```
Comment (moroshko) [inline package-lock.json:1]: Why remove `package-lock.json`?

File: package.json [hunk 3]
```diff
@@ -68,10 +75,17 @@
...
     "react-test-renderer": "16.14.0",
-    "rimraf": "3.0.2"
+    "rimraf": "3.0.2",
+    "rollup": "2.42.4",
+    "rollup-plugin-delete": "2.0.0",
+    "rollup-plugin-typescript2": "0.30.0",
+    "ts-jest": "26.5.4",
+    "tslib": "2.1.0",
+    "typescript": "4.2.3",
+    "yalc": "1.0.0-pre.50"
```
Comment (moroshko) [inline package.json:88]: This is funny how they never released a proper `1.0.0` 😄

File: rollup.config.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        targets: [
+          "dist/*",
+          //   "playground/src/component-lib",
+          "website/lib/basis-lib",
+        ],
+      }),
+      babel({ babelHelpers: "bundled" }),
+      commonjs(),
+      typescript({
+        include: ["src/*.js", "src/**/*.js", "src/*.ts(x)", "src/**/*.ts(x)"],
```
Comment (moroshko) [inline rollup.config.js:35]: Parens in `(x)` do nothing I believe. Also, shouldn't `x` be optional?
```
include: ["src/*.js", "src/**/*.js", "src/*.tsx?", "src/**/*.tsx?"],
```

File: rollup.config.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      //   banner: "/* eslint-disable */",
+      // },
+      { file: pkg.main, format: "cjs" },
+      { file: pkg.module, format: "esm" },
+    ],
+    plugins: [
+      del({
+        targets: [
+          "dist/*",
+          //   "playground/src/component-lib",
```
Comment (moroshko) [inline rollup.config.js:28]: Can we delete the comments above?

File: src/components/Container.js [hunk 6]
```diff
@@ -147,22 +140,22 @@ function Container(_props) {
...
+//   ...responsivePropType("flexGrow", PropTypes.bool),
+//   ...responsivePropType("hasBorder", PropTypes.bool),
+//   ...responsivePropType("textStyle", PropTypes.oneOf(TEXT_STYLES)),
+//   ...responsivePropType("textAlign", PropTypes.oneOf(TEXT_ALIGNS)),
+//   ...responsivePropType("overflow", PropTypes.string),
+//   ...responsivePropType("hasBreakpointWidth", PropTypes.bool),
+//   ...responsivePropType("hide", PropTypes.bool),
+//   children: PropTypes.node,
+//   testId: PropTypes.string,
+// };
```
Comment (moroshko) [inline src/components/Container.js:159]: Why did you remove all the PropTypes here? Wondering what's special about this `Container` component 🤔

File: src/components/Text.tsx [hunk 7]
```diff
@@ -121,30 +115,30 @@ function Text(props) {
...
+//   color: (props) => {
+//     allowedColors.forEach(({ textStyles, allowedColors }) => {
+//       if (
+//         textStyles.includes(props.textStyle) &&
+//         !allowedColors.includes(props.color)
+//       ) {
+//         return new Error(
+//           `Text: color="${props.color}" is not allowed for textStyle="${
+//             props.textStyle
+//           }". Must be one of: ${formatArray(allowedColors)}`
```
Comment (moroshko) [inline src/components/Text.tsx:132]: Would be good to find a TypeScript alternative for this warning.

**Review by moroshko — COMMENTED** [review]
Awesome work!

File: package.json [hunk 8]
```diff
@@ -1,21 +1,22 @@
...
+    "start-website": "cd website && npm run clean && npm start",
     "prettier": "prettier --write \"**/*.{js,json,css,md}\"",
     "lint": "eslint --max-warnings 0 \"**/*.js\"",
     "test": "cross-env BABEL_ENV=cjs jest",
     "percy": "percy snapshot public -c .percy.js",
     "prebuild-basis": "rimraf dist",
-    "build-basis:cjs": "cross-env NODE_ENV=production BABEL_ENV=cjs babel src -d dist --ignore \"src/**/*.test.js\"",
-    "build-basis:esm": "cross-env NODE_ENV=production BABEL_ENV=esm babel src -d dist/esm --ignore \"src/**/*.test.js\"",
-    "build-basis": "npm run build-basis:cjs && npm run build-basis:esm",
+    "build-basis": "rollup -c",
```
[the remark this answers is not in the document]
Comment (alifa20) [reply package.json:16]: -c is necessary , without this command,  rollup will fail

File: src/components/Container.js (hunk 6 above)
Comment (alifa20) [reply src/components/Container.js:159]: Container is referenced in in `src/components/Text.test.tsx` so we have to have this change to unblock this ticket.

File: src/components/Text.tsx (hunk 7 above)
Comment (alifa20) [reply src/components/Text.tsx:132]: I am keeping this proptypes for future prs to add types to this component

File: package-lock.json (hunk 2 above)
Comment (alifa20) [reply package-lock.json:1]: The package-lock.json is not removed.

File: package.json (hunk 3 above)
Comment (alifa20) [reply package.json:88]: 🙂

File: rollup.config.js (hunk 4 above)
Comment (alifa20) [reply rollup.config.js]: Parenthesis is needed for `npm run build-basis` to pass

File: package.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (moroshko) [reply package.json:16]: Why is it failing? `-c` is not mandatory (and having it without pointing to a specific config file makes even less sense to me). You can see in the [Quick start](https://rollupjs.org/guide/en/#quick-start) that `-c` is not used, so I wonder why do we need to specify it. What's the actual failure?

File: package.json [hunk 9]
```diff
@@ -32,26 +33,32 @@
     "react-keyed-flatten-children": "1.3.0"
   },
   "peerDependencies": {
-    "@emotion/core": "^10.0.28",
-    "prop-types": "^15.7.2",
-    "react": "^16.13.0",
-    "react-dom": "^16.13.0"
+    "@emotion/react": "11.1.5",
+    "prop-types": "15.7.2",
```
Comment (moroshko) [inline package.json:37]: Let's add `^` to all `peerDependencies`.

File: MIGRATING_TO_2.0.md [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# Migrating to V2
+
+- If you're using an older version of React, you'll have to upgrade to React 17.
+- You'll have to upgrade to Emotion 11. The main thing to notice is that you'll need to uninstall @emotion/core and install @emotion/react instead. More details [here ^](https://emotion.sh/docs/emotion-11){:target="\_blank"}
```
Comment (moroshko) [inline MIGRATING_TO_2.0.md:4]: Are you sure that this "target blank magic" works? It doesn't open in a new tab from this PR.

Comment (alifa20) [reply MIGRATING_TO_2.0.md]: Nah , will remove it.

File: README.md [hunk 11]
```diff
@@ -35,6 +35,8 @@ export default App;
 
 ## Developing locally
 
+Make sure nvm is installed.
```
Comment (moroshko) [inline README.md:38]: Let's remove this line.

File: package.json [hunk 12]
```diff
@@ -1,19 +1,20 @@
...
-    "lint": "eslint --max-warnings 0 \"**/*.js\"",
+    "lint": "eslint --max-warnings 0 --ext .js,.ts,.tsx .",
     "test": "cross-env BABEL_ENV=cjs jest",
     "percy": "percy snapshot public -c .percy.js",
     "prebuild-basis": "rimraf dist",
-    "build-basis:cjs": "cross-env NODE_ENV=production BABEL_ENV=cjs babel src -d dist --ignore \"src/**/*.test.js\"",
-    "build-basis:esm": "cross-env NODE_ENV=production BABEL_ENV=esm babel src -d dist/esm --ignore \"src/**/*.test.js\"",
-    "build-basis": "npm run build-basis:cjs && npm run build-basis:esm",
+    "build-basis": "rollup -c",
+    "build-watch": "rollup -c -w",
```
Comment (moroshko) [inline package.json:17]: You changed to dash here but above it's still using `:`.

**Review by moroshko — APPROVED** [review]
